Content.Server/Salvage/SalvageSystem.cs

```diff
@@ -296,125 +296,125 @@ private void SafeDeleteSalvage(EntityUid salvage)
             }
+            return false;
         }
 
-        private IEnumerable<SalvageMapPrototype> GetAllSalvageMaps() =>
-            _prototypeManager.EnumeratePrototypes<SalvageMapPrototype>();
-
-        private bool SpawnSalvage(SalvageMagnetComponent component)
+        private bool SpawnSalvage(EntityUid uid, SalvageMagnetComponent component)
         {
-            TryGetSalvagePlacementLocation(component, out var spl, out var spAngle);
+            var salvMap = _mapManager.CreateMap();
 
-            var forcedSalvage = _configurationManager.GetCVar(CCVars.SalvageForced);
-            List<SalvageMapPrototype> allSalvageMaps;
-            if (string.IsNullOrWhiteSpace(forcedSalvage))
+            Box2 bounds;
+            EntityUid? salvageEnt = null;
+            SalvageMapPrototype? salvageProto = null;
+            if (_random.Prob(component.AsteroidChance))
             {
-                allSalvageMaps = GetAllSalvageMaps().ToList();
+                var asteroidProto = _prototypeManager.Index<WeightedRandomPrototype>(component.AsteroidPool).Pick(_random);
+                salvageEnt = Spawn(asteroidProto, new MapCoordinates(0, 0, salvMap));
+                bounds = Comp<MapGridComponent>(salvageEnt.Value).LocalAABB;
             }
             else
             {
-                allSalvageMaps = new();
-                if (_prototypeManager.TryIndex<SalvageMapPrototype>(forcedSalvage, out var forcedMap))
-                {
-                    allSalvageMaps.Add(forcedMap);
-                }
-                else
-                {
-                    Logger.ErrorS("c.s.salvage", $"Unable to get forced salvage map prototype {forcedSalvage}");
-                }
+                var forcedSalvage = _configurationManager.GetCVar(CCVars.SalvageForced);
+                salvageProto = string.IsNullOrWhiteSpace(forcedSalvage)
+                    ? _random.Pick(_prototypeManager.EnumeratePrototypes<SalvageMapPrototype>().ToList())
+                    : _prototypeManager.Index<SalvageMapPrototype>(forcedSalvage);
+
+                bounds = salvageProto.Bounds;
             }
 
-            SalvageMapPrototype? map = null;
-            Vector2 spawnLocation = Vector2.Zero;
+            if (!TryGetSalvagePlacementLocation(uid, component, bounds, out var spawnLocation, out var spawnAngle))
+            {
+                Report(uid, component.SalvageChannel, "salvage-system-announcement-spawn-no-debris-available");
+                _mapManager.DeleteMap(salvMap);
+                return false;
+            }
 
-            for (var i = 0; i < allSalvageMaps.Count; i++)
+            if (salvageEnt is { } ent)
+            {
+                var salvXForm = Transform(ent);
+                _transform.SetParent(ent, salvXForm, _mapManager.GetMapEntityId(spawnLocation.MapId));
+                _transform.SetWorldPosition(salvXForm, spawnLocation.Position);
+                RaiseLocalEvent(ent, new LocalStructureLoadedEvent());
```

Comment:
sus amogus
just make LocalityLoaderSystem always load the object if there is no world controller.